node-wasm/src/subscriptions.rs

```diff
@@ -0,0 +1,250 @@
+pub struct IteratorResultObject {
+    /// Has the value true if the iterator is past the end of the
+    /// iterated sequence. In this case value optionally specifies
+    /// the return value of the iterator.
+    pub done: bool,
+    /// Any JavaScript value returned by the iterator.
+    /// Can be omitted when done is true.
+    pub value: JsValue,
+}
+
+impl IteratorResultObject {
+    pub fn done() -> Self {
+        Self {
+            done: true,
+            value: JsValue::UNDEFINED,
+        }
+    }
+
+    pub fn ready(value: JsValue) -> Self {
+        Self { done: false, value }
+    }
+}
+
+#[wasm_bindgen]
+#[derive(Clone)]
+pub struct AsyncIteratorImpl {
+    #[wasm_bindgen(skip)]
+    next: Rc<Closure<dyn FnMut() -> Promise>>,
+}
+
+impl AsyncIteratorImpl {
+    fn prepare_async_iterator_symbol(iterator: &JsValue) -> Result<()> {
+        let return_self_method = Reflect::get(iterator, &"return_self".into())?;
+        Reflect::set(iterator, &Symbol::async_iterator(), &return_self_method)?;
+        Ok(())
+    }
+}
+
+impl AsyncIteratorImpl {
+    pub fn into_raw(self) -> JsValue {
+        self.into()
+    }
+}
+
+#[wasm_bindgen]
+impl AsyncIteratorImpl {
+    pub fn next(&self) -> Promise {
+        Reflect::apply(
+            JsCast::unchecked_ref(self.next.as_ref().as_ref()),
+            &JsValue::UNDEFINED,
+            &js_sys::Array::new(),
+        )
+        .expect("apply on closure should succeed")
+        .unchecked_into()
+    }
+
+    pub fn return_self(self) -> JsValue {
+        JsValue::from(self.clone())
+    }
+}
+
+impl From<Closure<dyn FnMut() -> Promise>> for AsyncIteratorImpl {
+    fn from(next: Closure<dyn FnMut() -> Promise>) -> Self {
+        AsyncIteratorImpl {
+            next: Rc::new(next),
+        }
+    }
+}
+
+// Wrap provided port into SubscriptionStream and prepare it to be used as
+// js AsyncIterator. Assumes provided port was prepared with [`forward_stream_to_message_port`]
+pub(crate) fn into_async_iterator<S>(port: MessagePortLike) -> Result<AsyncIterator>
+where
+    S: DeserializeOwned + Into<JsValue> + 'static,
+{
+    let (feedback, receiver) = split_port(port)?;
+
+    let feedback = Rc::new(RefCell::new(feedback));
+    let receiver = Rc::new(RwLock::new(
+        receiver
+            .map(|result| {
+                from_value::<Result<S, JsSubscriptionError>>(result.data()).map_err(|e| {
+                    JsSubscriptionError {
+                        height: None,
+                        error: format!("error deserializing subscription item: {e}"),
+                    }
+                })?
+            })
+            .boxed_local(),
+    ));
+
+    let async_iterator: AsyncIteratorImpl = Closure::<dyn FnMut() -> Promise>::new(move || {
+        let (receiver, feedback) = (receiver.clone(), feedback.clone());
+        future_to_promise(async move {
+            if let Err(e) = feedback.borrow().send(&SubscriptionReceiverReady, &[]) {
```

Comment:
there's a comment on the `PortSender` to not use unit structs with it :detective: